compressors/external_compressors.py

```diff
@@ -0,0 +1,142 @@
+"""External compressors like gzip, etc. for testing/benchmarking purposes
+"""
+
+import os
+import tempfile
+from core.data_block import DataBlock
+from core.data_encoder_decoder import DataDecoder, DataEncoder
+from core.data_stream import TextFileDataStream, Uint8FileDataStream
+from core.encoded_stream import EncodedBlockReader, EncodedBlockWriter
+from core.prob_dist import ProbabilityDist
+from utils.bitarray_utils import BitArray, bitarray_to_uint, uint_to_bitarray
+from utils.test_utils import (
+    create_random_binary_file,
+    try_file_lossless_compression,
+    try_lossless_compression,
+)
+import zlib
+
+
+class ZlibExternalEncoder(DataEncoder):
+    def __init__(self, level=6):
+        self.level = level
+        # state stays alive across blocks so we can benefit
+        self.zlib_context = zlib.compressobj(level=self.level)
+        self.block_size_num_bits = 32  # num bits used to encode the block size at start
+
+    def reset(self):
+        # start new zlib context
+        self.zlib_context = zlib.compressobj(level=self.level)
+
+    def encode_block(self, data_block: DataBlock):
+        raw_bytes = bytes(data_block.data_list)
+
+        # flush below with Z_SYNC_FLUSH that ensures decompress is able to decompress the
+        # data till now. Note that this still utilizes this block for finding matches when
+        # we are compressing the next block (as opposed to Z_FULL_FLUSH that resets the state).
+        # See https://www.zlib.net/manual.html for more information
+        compressed_bytes = self.zlib_context.compress(raw_bytes) + self.zlib_context.flush(
+            zlib.Z_SYNC_FLUSH
+        )
+
+        # might be inefficient to convert to BitArray since it will be later be
+        # converted back to bytes when writing to file
+        # also, should we worry about endianness?
+
+        # at start write the compressed size because zlib decoder expects to get complete blocks
+        # and cannot determine when we want to end
+        compressed_bitarray = BitArray(
+            uint_to_bitarray(len(compressed_bytes) * 8, bit_width=self.block_size_num_bits)
+        )
+        compressed_bitarray.frombytes(compressed_bytes)
```

Comment:
why are we setting compressed_bitarray twice?

Reply:
First one just puts the length in there. Second one actual appends the data to it.